keylime/ima_ast.py

```diff
@@ -0,0 +1,320 @@
+            return struct.pack(f"{len(self.name)}sB{TCG_EVENT_NAME_LEN_MAX - len(self.name)}s",
+                               self.name.encode("utf-8"),
+                               NULL_BYTE,
+                               bytearray(TCG_EVENT_NAME_LEN_MAX - len(self.name)))
+
+        return struct.pack(f"<I{len(self.name)}sB",
+                           len(self.name) + 1,
+                           self.name.encode("utf-8"),
+                           NULL_BYTE)
+
+
+class Digest:
+    """
+    Class for types "d" and "d-ng" with and without algorithm
+    """
+    hash: bytes
+    algorithm: Optional[str] = None
+    legacy: bool = False
+
+    def __init__(self, digest: str, legacy=False):
+        self.legacy = legacy
+        tokens = digest.split(":")
+        if len(tokens) == 1:
+            try:
+                self.hash = codecs.decode(tokens[0].encode("utf-8"), "hex")
+            except binascii.Error:
+                raise ParserError(f"Digest hash is not valid hex. Got: {tokens[0]}")
+            if not (len(self.hash) == SHA_DIGEST_LEN or len(self.hash) == MD5_DIGEST_LEN):
+                raise ParserError(
+                    "Cannot create Digest. No hash algorithm is provided and hash does not belong to a md5 or sha1 hash.")
+        elif len(tokens) == 2:
+            try:
+                self.hash = codecs.decode(tokens[1].encode("utf-8"), "hex")
+            except binascii.Error:
+                raise ParserError(f"Digest hash is not valid hex. Got: {tokens[1]}")
+            self.algorithm = tokens[0]
+        else:
+            raise ParserError(f"Cannot create Digest expected 1 or 2 tokens got: {len(tokens)} for {digest}")
+
+    def struct(self):
+        # The legacy format "d" has fixed length, so it does not contain a length attribute
+        if self.legacy:
+            return struct.pack(f"<{len(self.hash)}s", self.hash)
+
+        if self.algorithm is None:
+            return struct.pack(f"<I{len(self.hash)}s", len(self.hash), self.hash)
+        # After the ':' must be a '\O':
+        # https://elixir.bootlin.com/linux/v5.12.10/source/security/integrity/ima/ima_template_lib.c#L230
+        return struct.pack(f"<I{len(self.algorithm)}sBB{len(self.hash)}s",
+                           len(self.algorithm) + 2 + len(self.hash),
+                           self.algorithm.encode("utf-8"),
+                           COLON_BYTE,
+                           NULL_BYTE,
+                           self.hash)
+
+
+class Ima(Mode):
+    """
+    Class for "ima". Contains the digest and a path.
+    """
+    digest: Digest
+    path: Name
+
+    def __init__(self, data: str):
+        tokens = data.split()
+        if len(tokens) != 2:
+            raise ParserError()
+        self.digest = Digest(tokens[0], legacy=True)
+        self.path = Name(tokens[1], legacy=True)
+
+    def hash(self):
+        tohash = self.digest.struct() + self.path.struct()
+        return hashlib.sha1(tohash).digest()
```

Comment:
What is it hashing in this case with a sha1?

Reply:
The hash function reconstructs the template hash that is in the IMA log.
In this case the template `ima` is build from a `d` (digest) and  `n` (name). 
To match the template hash, get the byte representation of the digest and name with `struct()`, concat them and then hash the result.

Reply:
Please 'd' and 'n' fields should really be limited to the original "ima" template.   At minimum use the ima-ng template, which are based on the 'd-ng' and 'n-ng' fields.  Using the 'ima-sig' template would be better.

Reply:
"The old code skipped the validation of the template hash if it none or all bits set. Why is that the case?
I couldn't find a good explanation by looking around the kernel IMA implementation and Keylime git history.
If this behavior is necessary it should be documented before merging."

IMA support for updating multiple TPM 2.0 banks based on the per bank crypto algorithm was relatively recently upstreamed.   The second field of the IMA measurement list still contains a SHA1 hash of the template data.  To validate the other TPM banks (e.g. sha256), requires walking the IMA measurement verifying the sha1 template data hash matches the template data, before re-calculating and extending the template data hash for each TPM bank enabled.

Reply:
The `d` and `n` field are limited to the `ima` template. We treat `d`/`d-ng` or `n`/`n-ng` as the same object for validators to consume but reconstructing their hashes differently. 

In the future we should add support for also returning the sh256 etc. hashes and check them also. The issue that is here: https://github.com/keylime/keylime/issues/30.
Does IMA recalculate the hash for each algorithm or does it hash with sha1 and than adds padding for the other algorithms?

Reply:
> The `d` and `n` field are limited to the `ima` template. We treat `d`/`d-ng` or `n`/`n-ng` as the same object for validators to consume but reconstructing their hashes differently.
> 
> In the future we should add support for also returning the sh256 etc. hashes and check them also. The issue that is here: #30.
> Does IMA recalculate the hash for each algorithm or does it hash with sha1 and than adds padding for the other algorithms?

Support for calculating and extending per TPM bank digests was upstreamed in Linux 5.8.  Refer to the upstream kernel commit 1ea973df6e21 ("ima: Calculate and extend PCR with digests in ima_template_entry"). - Mimi